### PR TITLE
Feat implement whatsapp with vonage bun

### PR DESCRIPTION
**Pull request by parthj000**
hey @gewenyu99 @Haimantika @loks0n , i have raised the PR please review it !!

This is a video demonstrating the bun whatsapp bot with vonagee-api

[](url)

https://github.com/appwrite/templates/assets/146522413/162ff112-5585-475c-b7d6-c0a725e21f19

### CODE REVIEW

**Comment by parthj000** [comment]
image

File: bun/starter/whatsapp-with-vonage/README.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+`VONAGE_API_KEY`
+
+`VONAGE_ACCOUNT_SECRET`
+
+`VONAGE_SIGNATURE_SECRET`
+
+`VONAGE_WHATSAPP_NUMBER`
+
+`PORT`
+## Run Locally
```
Comment (gewenyu99) [inline bun/starter/whatsapp-with-vonage/README.md:125]: Please remove these additional sections and keep this consistent with the existing Whatsapp vonage templates

File: bun/starter/whatsapp-with-vonage/src/main.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
+import express from "express"
```
Comment (gewenyu99) [inline bun/starter/whatsapp-with-vonage/src/main.js:1]: Why are you importing express? You don't need express with Appwrite functions, it's already got a node based webserver at the front

**Review by gewenyu99 — CHANGES_REQUESTED** [review]
Hey, I see a few issues.
1. you've nested this under bun/starter, which is not entirely correct
2. You've created a standalone express server, this should be an [Appwrite Function](https://appwrite.io/docs/products/functions)

Thanks!

**Comment by parthj000** [comment]
@loks0n i have done all the required changes !!

**Comment by gewenyu99** [comment]
Hey,

Due to time constraints, I'm going to mark this PR `hacktoberfest-accepted` for now so you get DO's Hacktoberfest rewards. We'll continue to work with you on this issue for review and merge.

When it is merged, we'll contact you for Appwrite-specific Hacktoberfest swag.

Thanks for helping us improve Appwrite!

File: bun/starter/src/main.ts [hunk 3]
```diff
@@ -25,9 +25,9 @@ export default async ({ req, res, log, error }: any) => {
 
   // `res.json()` is a handy helper for sending JSON
   return res.json({
-    motto: "Build Fast. Scale Big. All in One Place.",
+    motto: "Build like a team of hundreds_",
```
Comment (loks0n) [inline bun/starter/src/main.ts:28]: Please revert this change.

File: bun/starter/package.json
Comment (loks0n) [inline bun/starter/package.json:1]: Please revert this change.

File: bun/starter/bun.lockb
Comment (loks0n) [inline bun/starter/bun.lockb:1]: Please revert this change.

File: bun/starter/README.md
Comment (loks0n) [inline bun/starter/README.md:1]: Please revert this change.

File: bun/starter/.gitignore
Comment (loks0n) [inline bun/starter/.gitignore:1]: Please revert this change.

File: bun/whatsapp-with-vonage/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+## POST/
+
+Webhooks sent by vonage to update about the status of message
+
+| Name          | Description                        | Location | Type                | Sample Value         |
+| ------------- | ---------------------------------- | -------- | ------------------- | -------------------- |
+| Status  | showing the status of message    | body  | string| submitted/read/delievered          |
+
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:47]: Can you remove all the double empty lines in this file?

File: bun/whatsapp-with-vonage/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+}
+```
+
+## POST/
+
+Webhooks sent by vonage to update about the status of message
+
+| Name          | Description                        | Location | Type                | Sample Value         |
+| ------------- | ---------------------------------- | -------- | ------------------- | -------------------- |
+| Status  | showing the status of message    | body  | string| submitted/read/delievered          |
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:45]: Please format the file with prettier

File: bun/whatsapp-with-vonage/README.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+| Sample Value  | `ZACYi3...ZXGjsch`                                                                                                |
+
+## CONFIGURATION
+
+
+| Setting           | Value                    |
+| ----------------- | ------------------------ |
+| Runtime           | Bun (1.0)              |
+| Entrypoint        | `src/main.js`            |
+| Build Commands    | `Bun install src/main.js` |
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:106]: This is not correct, please see the bun starter as an example

File: bun/whatsapp-with-vonage/README.md [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+| Required      | Yes                                                                                                            |
+| Sample Value  | `ZACYi3...ZXGjsch`                                                                                                |
+
+## CONFIGURATION
+
+
+| Setting           | Value                    |
+| ----------------- | ------------------------ |
+| Runtime           | Bun (1.0)              |
+| Entrypoint        | `src/main.js`            |
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:105]: This is not correct, please see the bun starter as an example

File: bun/whatsapp-with-vonage/README.md [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+`VONAGE_ACCOUNT_SECRET`
+
+`VONAGE_SIGNATURE_SECRET`
+
+`VONAGE_WHATSAPP_NUMBER`
+
+## Documentation
+
+[Vonage Documentation ](https://developer.vonage.com/en/documentation)
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:126]: We already mention this above, please remove this

File: bun/whatsapp-with-vonage/public/index.html
Comment (loks0n) [inline bun/whatsapp-with-vonage/public/index.html:1]: Can you use the HTML from the `node/whatsapp-with-vonage` function and put it in `static` folder please

File: bun/whatsapp-with-vonage/src/main.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import jwt from 'jsonwebtoken';
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.js:2]: There are loads of random empty lines in the this file... please remove them.

File: bun/whatsapp-with-vonage/src/main.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+export default async ({ req, res, log, error }) => {
+
+    const html = file("public/index.html")
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.js:6]: This won't work, please check out how the other bun templates are doing it.

File: bun/whatsapp-with-vonage/src/main.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import jwt from 'jsonwebtoken';
+
+
+export default async ({ req, res, log, error }) => {
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.js:4]: Try structuring the code in a similar way to the node template.

File: bun/whatsapp-with-vonage/src/main.js [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        });
+
+        const requestOptions = {
+            method: 'POST',
+            headers: myHeaders,
+            body: raw,
+
+        };
+        if (!(req.body.text == null)) {
+            await fetch("https://messages-sandbox.nexmo.com/v1/messages", requestOptions)
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.js:56]: You need to import fetch from `undici`

File: bun/whatsapp-with-vonage/src/main.js [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+                    error: "can't verify"
+
+                }, 400);
+            }
+
+
+        })
+
+        const myHeaders = new Headers();
+        myHeaders.append("Accept", "application/json");
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.js:37]: This header is not required

File: bun/whatsapp-with-vonage/src/main.js [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+        myHeaders.append("Content-Type", "application/json");
+        myHeaders.append("Authorization", `Basic ${basicAuthToken}`);
+
+        const raw = JSON.stringify({
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.js:41]: raw is not a very descriptive name

**Review by loks0n — CHANGES_REQUESTED** [review]

**Comment by parthj000** [comment]
@loks0n , i have done the required changes !

File: bun/starter/README.md [hunk 15]
```diff
@@ -26,7 +26,7 @@ Sample `200` Response:
 
 ```json
 {
-  "motto": "Build Fast. Scale Big. All in One Place.",
+  "motto": "Build like a team of hundreds_",
```
Comment (loks0n) [inline bun/starter/README.md:29]: Please revert this change

File: bun/starter/.gitignore
Comment (loks0n) [inline bun/starter/.gitignore:1]: Revert this change

File: bun/starter/README.md [hunk 16]
```diff
@@ -45,4 +45,4 @@ Sample `200` Response:
 
 ## 🔒 Environment Variables
 
-No environment variables required.
+No environment variables required.
```
Comment (loks0n) [inline bun/starter/README.md:48]: Please revert this change

File: bun/starter/package.json [hunk 17]
```diff
@@ -13,4 +13,4 @@
   "devDependencies": {
     "prettier": "^3.0.0"
   }
-}
+}
```
Comment (loks0n) [inline bun/starter/package.json:16]: Please revert this change

File: bun/starter/bun.lockb
Comment (loks0n) [inline bun/starter/bun.lockb:1]: Please revert this change

File: bun/starter/src/main.ts [hunk 18]
```diff
@@ -1,4 +1,4 @@
-import { Client } from 'node-appwrite';
+import { Client } from "node-appwrite";
```
Comment (loks0n) [inline bun/starter/src/main.ts:1]: Please revert this change

File: bun/whatsapp-with-vonage/README.md [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+| Required     | Yes                |
+| Sample Value | `ZACYi3...ZXGjsch` |
+
+## CONFIGURATION
+
+| Setting           | Value          |
+| ----------------- | -------------- |
+| Runtime           | Bun (1.0)      |
+| Entrypoint        | `src/main.ts`  |
+| Build Commands    | `Bun install ` |
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:100]: Should be lowercase `bun install` here

File: bun/whatsapp-with-vonage/src/main.ts [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+  if (req.method === "GET") {
+    const fileName = Bun.resolveSync("../static/index.html", import.meta.dir);
+    const html = await Bun.file(fileName).text();
+
+    return res.send(html, 200, {
+      "Content-Type": "text/html; charset=utf-8",
+    });
+  }
+
+  if (req.method === "POST") {
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:14]: This if state is not required

File: bun/whatsapp-with-vonage/src/main.ts [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+      channel: "whatsapp",
+    });
+
+    const requestOptions = {
+      method: "POST",
+      headers: myHeaders,
+      body: bodyContent,
+    };
+    if (!(req.body.text == null)) {
+      await fetch(
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:53]: Can we just do all this inline:
```js
  await fetch(`https://messages-sandbox.nexmo.com/v1/messages`, {
    method: 'POST',
    body: JSON.stringify({
      from: process.env.VONAGE_WHATSAPP_NUMBER,
      to: req.body.from,
      message_type: 'text',
      text: `Hi there! You sent me: ${req.body.text}`,
      channel: 'whatsapp',
    }),
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Basic ${basicAuthToken}`,
    },
  });
```

File: bun/whatsapp-with-vonage/src/main.ts [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+          return res.json(
+            {
+              ok: false,
+              error: "can't verify",
+            },
+            400
+          );
+        }
+      }
+    );
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:33]: This is quite complicated. Take a look the [node template](https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/node/whatsapp-with-vonage/src/main.js#L20)
```js
  const token = (req.headers.authorization ?? '').split(' ')[1];
  var decoded = jwt.verify(token, process.env.VONAGE_API_SIGNATURE_SECRET, {
    algorithms: ['HS256'],
  });

  if (sha256(req.bodyRaw) != decoded['payload_hash']) {
    return res.json({ ok: false, error: 'Payload hash mismatch.' }, 401);
  }
```

**Review by loks0n — CHANGES_REQUESTED** [review]

**Comment by parthj000** [comment]
@loks0n 
I m little confused which changes you want me to revert .
Besides it i have added payload hash verification , and also added the asked changes, can you review them !

File: bun/starter/README.md (hunk 16 above)
Comment (loks0n) [reply bun/starter/README.md:47]: Reset this file please, there is a whitespace change`git checkout origin/main bun/starter/README.md`

File: bun/starter/bun.lockb
Comment (loks0n) [inline bun/starter/bun.lockb:1]: Again reset this file

File: bun/starter/package.json
Comment (loks0n) [reply bun/starter/package.json:1]: Reset this file please, there is a whitespace changegit checkout origin/main bun/starter/README.md

File: bun/whatsapp-with-vonage/src/main.ts [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+
+export default async ({ req, res, log, error }) => {
+  if (req.method === "GET") {
+    const fileName = Bun.resolveSync("../static/index.html", import.meta.dir);
+    const html = await Bun.file(fileName).text();
+
+    return res.send(html, 200, {
+      "Content-Type": "text/html; charset=utf-8",
+    });
+  }
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:12]: We use a utils file for this is other templates, please look

File: bun/whatsapp-with-vonage/src/main.ts [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+        ok: false,
+        error: "Payload mismatched",
+      });
+    }
+  } catch (err) {
+    return res.json({
+      ok: false,
+      error: "can't verify",
+    });
+  }
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:38]: Please use the same flow as the existing node template. This is completely different

File: bun/whatsapp-with-vonage/src/main.ts [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+        message_type: "text",
+        text: `You sent me: ${req.body.text}`,
+        channel: "whatsapp",
+      }),
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: `Basic ${basicAuthToken}`,
+      },
+    });
+  }
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:58]: Swap the case round and use early return here please https://refactoring.guru/replace-nested-conditional-with-guard-clauses

**Review by loks0n — CHANGES_REQUESTED** [review]

**Comment by loks0n** [comment]
> @loks0n I m little confused which changes you want me to revert . Besides it i have added payload hash verification , and also added the asked changes, can you review them !

Please follow the Node template as closely as possible.

File: bun/whatsapp-with-vonage/src/main.ts [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+      Authorization: `Basic ${basicAuthToken}`,
+    },
+  });
+
+  return res.json(
+    {
+      ok: true,
+    },
+    200
+  );
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:64]: Formatting looks weird, please run prettier

File: bun/whatsapp-with-vonage/src/main.ts [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+    return res.json(
+      {
+        ok: false,
+        error: "Payload hash mismatched",
+      },
+      401
+    );
+  }
+
+  const secret = `${Bun.env.VONAGE_API_KEY}:${Bun.env.VONAGE_ACCOUNT_SECRET}`;
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:31]: This variable is unnecessary please match https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/node/whatsapp-with-vonage/src/main.js#L35

File: bun/whatsapp-with-vonage/src/main.ts [hunk 28]
```diff
@@ -0,0 +1,65 @@
+import jwt from "jsonwebtoken";
+import { fetch } from "undici";
+import { getStaticFiles } from "./utils";
+
+export default async ({ req, res, log, error }) => {
+  if (req.method === "GET") {
+    return res.send(getStaticFiles, 200, {
+      "Content-Type": "text/html; charset=utf-8",
+    });
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:9]: Why is it called `getStaticFiles` when it returns the contents of a single file?

Please make a function of format like `getStaticFile(filename: string)` as like https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/node/whatsapp-with-vonage/src/main.js#L15

File: bun/whatsapp-with-vonage/src/main.ts [hunk 29]
```diff
@@ -0,0 +1,65 @@
+import jwt from "jsonwebtoken";
+import { fetch } from "undici";
+import { getStaticFiles } from "./utils";
+
+export default async ({ req, res, log, error }) => {
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:5]: We need to throw if the environment variables are not set like here https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/node/whatsapp-with-vonage/src/main.js#L7

File: bun/whatsapp-with-vonage/src/main.ts [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+
+  if (hashedValue != decodedToken["payload_hash"]) {
+    return res.json(
+      {
+        ok: false,
+        error: "Payload hash mismatched",
+      },
+      401
+    );
+  }
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:29]: This is very different to the node template, please match the node template as closely as possible but using Bun utilites.
https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/node/whatsapp-with-vonage/src/main.js#L20

File: bun/whatsapp-with-vonage/src/utils.ts [hunk 31]
```diff
@@ -0,0 +1,5 @@
+export async function getStaticFiles() {
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/utils.ts:1]: Please copy other bun templates to get the best version of this function

**Review by loks0n — CHANGES_REQUESTED** [review]
Hi, 

I'm spending a lot of time going back and forth with this PR - please try to address all my comments before resubmitting for review.

File: bun/whatsapp-with-vonage/src/utils.ts [hunk 32]
```diff
@@ -0,0 +1,21 @@
+//  Returns the contents of a file in the static folder
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/utils.ts:2]: Please change this comment to match the other TS templates https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/deno/whatsapp-with-vonage/src/utils.ts#L2

File: bun/whatsapp-with-vonage/src/utils.ts [hunk 33]
```diff
@@ -0,0 +1,21 @@
...
+export async function getStaticFiles(fileName: string) {
+  const filePath = Bun.resolveSync(`../static/${fileName}`, import.meta.dir);
+  const html = await Bun.file(filePath).text();
+  return html;
+}
+
+// Throws an error if any of the keys are missing from the object
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/utils.ts:9]: https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/deno/whatsapp-with-vonage/src/utils.ts#L2

File: bun/whatsapp-with-vonage/src/main.ts [hunk 34]
```diff
@@ -0,0 +1,66 @@
+import jwt from "jsonwebtoken";
+import { fetch } from "undici";
+import { getStaticFiles, throwIfMissing } from "./utils.ts";
+
+export default async ({ req, res, log, error }: any) => {
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:5]: Can we add basic types here as shown here: https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/deno/whatsapp-with-vonage/src/main.ts#L6C1-L12C1

File: bun/whatsapp-with-vonage/README.md [hunk 35]
```diff
@@ -0,0 +1,111 @@
...
+| Name   | Description                   | Location | Type   | Sample Value              |
+| ------ | ----------------------------- | -------- | ------ | ------------------------- |
+| Status | showing the status of message | body     | string | submitted/read/delievered |
+
+**Response:**
+
+```json
+{
+  "ok": 200,
+  "status": "delievered/submitted/read"
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:63]: This is not what you actually return.

File: bun/whatsapp-with-vonage/src/main.ts [hunk 36]
```diff
@@ -0,0 +1,66 @@
...
+    return res.json({ ok: false, error: err.message }, 400);
+  }
+
+  const basicAuthToken: string = btoa(
+    `${Bun.env.VONAGE_API_KEY}:${Bun.env.VONAGE_ACCOUNT_SECRET}`
+  );
+
+  if (req.body.text == null) {
+    return res.json({ ok: true, status: req.body.status }, 200);
+  }
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:48]: This is not necessary, please remove this

File: bun/whatsapp-with-vonage/src/main.ts [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+
+  if (req.method === "GET") {
+    return res.send(getStaticFiles("index.html"), 200, {
+      "Content-Type": "text/html; charset=utf-8",
+    });
+  }
+  const authHeader: string = (req.headers.authorization ?? "").split(" ")[1];
+  const decodedToken = jwt.verify(authHeader, Bun.env.VONAGE_SIGNATURE_SECRET, {
+    algorithms: ["HS256"],
+  });
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:21]: Please add a space before this section and use structure similar to here
https://github.com/appwrite/templates/blob/6dab35f1f221ba075640d93f5a8e374f4fba4349/node/whatsapp-with-vonage/src/main.js#L20

First assign `token` variable

Then assign `decoded` variable

File: bun/whatsapp-with-vonage/src/main.ts [hunk 38]
```diff
@@ -0,0 +1,66 @@
...
+
+  try {
+    throwIfMissing(decodedToken, ["payload_hash"]);
+  } catch (err) {
+    return res.json({ ok: false, error: err.message }, 400);
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:26]: This error is undocumented in the README

**Review by loks0n — CHANGES_REQUESTED** [review]

**Comment by parthj000** [comment]
@loks0n , but what will happen to the event webhook ,If I remove this then incoming event webhooks will not be addressed and in documentation it is written that the webhook must be responded with 2XX response and if not done the request will be retried ... so will it not cause any problem and is it efficient to not account the webhooks giving us the status of message ?
And I have removed "text" as a parameter from the throwIfMissing function .

Documentation -https://developer.vonage.com/en/api/messages-olympus#websockets

**Comment by parthj000** [comment]
Please review this ! @loks0n

File: bun/whatsapp-with-vonage/README.md [hunk 39]
```diff
@@ -0,0 +1,118 @@
...
+  "error": "Missing required parameter: from"
+}
+```
+
+```json
+{
+  "ok": false,
+  "error": "Missing required parameter: payload_hash"
+}
+```
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:46]: Let's just remove this example to match the other readmes

File: bun/whatsapp-with-vonage/README.md [hunk 40]
```diff
@@ -0,0 +1,118 @@
...
+
+Simple bot to answer WhatsApp messages.
+
+## 🧰 Usage
+
+## GET/
+
+HTML form interacting with the function.
+
+## POST/
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:11]: Please add a space between POST and /

File: bun/whatsapp-with-vonage/README.md [hunk 41]
```diff
@@ -0,0 +1,118 @@
+## 💬 Bun Whatsapp bot with Vonage Function
+
+Simple bot to answer WhatsApp messages.
+
+## 🧰 Usage
+
+## GET/
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:7]: Please add a space between POST and /

File: bun/whatsapp-with-vonage/README.md [hunk 42]
```diff
@@ -0,0 +1,118 @@
...
+| Status | showing the status of message | body     | string | submitted/read/delievered |
+
+**Response:**
+
+```json
+{
+  "ok": 200,
+  "status": "delievered"
+}
+```
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/README.md:72]: Let's remove this section. We have already covered it.

File: bun/whatsapp-with-vonage/src/main.ts [hunk 43]
```diff
@@ -0,0 +1,81 @@
...
+
+type Context = {
+  req: any;
+  res: any;
+  log: (msg: any) => void;
+  error: (msg: any) => void;
+};
+
+export default async ({ req, res, log, error }: Context) => {
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:13]: Please remove the newline here

File: bun/whatsapp-with-vonage/src/main.ts [hunk 44]
```diff
@@ -0,0 +1,81 @@
...
+export default async ({ req, res, log, error }: Context) => {
+  
+  throwIfMissing(Bun.env, [
+    "VONAGE_API_KEY",
+    "VONAGE_ACCOUNT_SECRET",
+    "VONAGE_WHATSAPP_NUMBER",
+    "VONAGE_SIGNATURE_SECRET",
+  ]);
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:20]: Please remove the double newline here and replace with a single newline

File: bun/whatsapp-with-vonage/src/main.ts [hunk 45]
```diff
@@ -0,0 +1,81 @@
...
+    return res.send(getStaticFiles("index.html"), 200, {
+      "Content-Type": "text/html; charset=utf-8",
+    });
+  }
+
+  const token: string = (req.headers.authorization ?? "").split(" ")[1];
+  const decodedToken = jwt.verify(token, Bun.env.VONAGE_SIGNATURE_SECRET, {
+    algorithms: ["HS256"],
+  });
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:32]: Please remove the double newline here, and replace with a single new line

File: bun/whatsapp-with-vonage/src/main.ts [hunk 46]
```diff
@@ -0,0 +1,81 @@
...
+  try {
+    throwIfMissing(decodedToken, ["payload_hash"]);
+  } catch (err) {
+    return res.json({ ok: false, error: err.message }, 400);
+  }
+  
+  const hasher = new Bun.CryptoHasher("sha256");
+  const hashedValue = hasher.update(req.bodyRaw).digest("hex").toString();
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:42]: Please remove the double newline here and replace with a single newline

File: bun/whatsapp-with-vonage/src/main.ts [hunk 47]
```diff
@@ -0,0 +1,81 @@
...
+  }
+  
+  const hasher = new Bun.CryptoHasher("sha256");
+  const hashedValue = hasher.update(req.bodyRaw).digest("hex").toString();
+  
+
+  if (hashedValue != decodedToken["payload_hash"]) {
+    return res.json({ ok: false, error: "Payload hash mismatched" }, 401);
+  }
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:47]: Please remove the double newline here and replace with a single newline

File: bun/whatsapp-with-vonage/src/main.ts [hunk 48]
```diff
@@ -0,0 +1,81 @@
...
+    return res.json({ ok: false, error: "Payload hash mismatched" }, 401);
+  }
+  
+
+  try {
+    throwIfMissing(req.body, ["from", "to"]);
+  } catch (err) {
+    return res.json({ ok: false, error: err.message }, 400);
+  }
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:54]: Please remove the double newline here and replace with a single newline

File: bun/whatsapp-with-vonage/src/main.ts [hunk 49]
```diff
@@ -0,0 +1,81 @@
...
+    throwIfMissing(req.body, ["from", "to"]);
+  } catch (err) {
+    return res.json({ ok: false, error: err.message }, 400);
+  }
+  
+
+  if (req.body.text == null) {
+    return res.json({ ok: true, status: req.body.status }, 200);
+  }
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:59]: Please remove the double newline here and replace with a single newline

File: bun/whatsapp-with-vonage/src/utils.ts [hunk 50]
```diff
@@ -0,0 +1,25 @@
...
+  if (missing.length > 0) {
+    throw new Error(`Missing required parameter:${missing.join(", ")}`);
+  }
+}
+
+
+/*
+ * Returns the contents of a file in the static folder
+ */
+
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/utils.ts:20]: Please remove the newline here

File: bun/whatsapp-with-vonage/static/index.html [hunk 51]
```diff
@@ -0,0 +1,34 @@
+<!doctype html >
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/static/index.html:1]: Please remove the space here <!doctype html >
```suggestion
<!doctype html>
```

File: bun/whatsapp-with-vonage/src/utils.ts [hunk 52]
```diff
@@ -0,0 +1,25 @@
...
+export async function getStaticFiles(fileName: string) {
+  const filePath = Bun.resolveSync(`../static/${fileName}`, import.meta.dir);
+  const html = await Bun.file(filePath).text();
+  return html;
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/utils.ts:24]: Let's remove the inline variable and just `return await Bun.file(filePath).text()`

File: bun/whatsapp-with-vonage/src/utils.ts [hunk 53]
```diff
@@ -0,0 +1,25 @@
...
+    throw new Error(`Missing required parameter:${missing.join(", ")}`);
+  }
+}
+
+
+/*
+ * Returns the contents of a file in the static folder
+ */
+
+export async function getStaticFiles(fileName: string) {
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/utils.ts:21]: Can we rename this `getStaticFile`

**Review by loks0n — CHANGES_REQUESTED** [review]

**Comment by parthj000** [comment]
updated !

File: bun/whatsapp-with-vonage/src/main.ts [hunk 54]
```diff
@@ -0,0 +1,74 @@
...
+  try {
+    throwIfMissing(req.body, ["from", "to"]);
+  } catch (err) {
+    return res.json({ ok: false, error: err.message }, 400);
+  }
+
+  if (req.body.text == null) {
+    return res.json({ ok: true, status: req.body.status }, 200);
+  }
```
Comment (loks0n) [inline bun/whatsapp-with-vonage/src/main.ts:52]: Please remove this.

**Review by loks0n — CHANGES_REQUESTED** [review]

**Review by loks0n — APPROVED** [review]

**Comment by gewenyu99** [comment]
Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship. 

Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.

**Comment by parthj000** [comment]
@gewenyu99 ,
pars0942

> Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship.
> 
> Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.

**Comment by parthj000** [comment]
Hey , @gewenyu99,  I have not been contacted for the swags yet :)

**Comment by adityaoberai** [comment]
> @gewenyu99 , pars0942
> 
> > Hey there! There were a lot of big PRs during this Hacktoberfest, and we wanted to give everyone ample time to collaborate with our engineering team. If you were able to merge your PRs during October, amazing. If it’s still not merged, don’t worry about it either. Either way, we’ve got your Hacktoberfest swag minted and ready to ship.
> > Please comment with your Discord username here so we can contact you about your shipping information to deliver your Hacktoberfest swag.

Hey Parth
Just sent you a friend request on DIscord. If you can accept it and DM me, I'll send over the form.

**Pull request merged**